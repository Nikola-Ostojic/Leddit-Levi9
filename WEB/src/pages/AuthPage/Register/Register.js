import React from "react";
import { Button, TextField } from "@material-ui/core";

export class Register extends React.Component {
  state = {
    username: "",
    email: "",
    password: "",
    confirmPassword: "",
    validationError: null
  };

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  register = (e) => {
    e.preventDefault();
    const { username, email, password, confirmPassword } = this.state;
    const { register } = this.props;
    if (password !== confirmPassword) {
      this.setState({
        validationError: "Password and confirm password don't match"
      });
      return;
    }
    register(username, email, password, confirmPassword);
  };
  render() {
    const {
      username,
      email,
      password,
      confirmPassword,
      validationError
    } = this.state;
    const { error } = this.props;

    return (
      <div>
        <h3>Register</h3>
        <form onSubmit={this.register} method="POST">
          <TextField
            fullWidth
            margin="normal"
            name="username"
            label="Username"
            type="text"
            required
            onChange={this.handleChange}
            value={username}
          />
          <TextField
            fullWidth
            margin="normal"
            name="email"
            label="Email"
            type="email"
            required
            onChange={this.handleChange}
            value={email}
          />
          <TextField
            fullWidth
            margin="normal"
            name="password"
            type="password"
            label="Password"
            required
            onChange={this.handleChange}
            value={password}
          />
          <TextField
            fullWidth
            margin="normal"
            name="confirmPassword"
            type="password"
            label="Confirm password"
            required
            onChange={this.handleChange}
            value={confirmPassword}
          />
          <br />
          <br />
          <Button variant="contained" color="primary" type="submit">
            Register
          </Button>
        </form>
        {error && error}
        {validationError && validationError}
      </div>
    );
  }
}

export default Register;
