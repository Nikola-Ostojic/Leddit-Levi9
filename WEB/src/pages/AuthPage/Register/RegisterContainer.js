import { connect } from "react-redux";

import { getError } from "modules/Auth/AuthSelectors";
import { register } from "modules/Auth/AuthActions";

import Register from "./Register";

export const mapStateToProps = (state) => ({
  error: getError(state)
});

export default connect(mapStateToProps, { register })(Register);
