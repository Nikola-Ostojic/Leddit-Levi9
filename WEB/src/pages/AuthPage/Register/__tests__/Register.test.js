import React from "react";
import { shallow } from "enzyme";
import { TextField } from "@material-ui/core";

import { Register } from "../Register";

describe("Register component tests", () => {
  it("Should match snapshot", () => {
    const wrapper = shallow(<Register />);

    expect(wrapper).toMatchSnapshot();
  });

  it("Should call register if the required fields are filled out in the form", () => {
    // Arrange
    const register = jest.fn();
    const wrapper = shallow(<Register register={register} />);
    const username = "Test";
    const email = "Test@gmail.com";
    const password = "Test";
    const confirmPassword = "Test";

    // Act
    wrapper.setState({ username, email, password, confirmPassword });
    const form = wrapper.find("form");
    form.simulate("submit", {
      preventDefault: jest.fn()
    });

    // Assert
    expect(register).toBeCalledWith(username, email, password, confirmPassword);
  });

  it("Should not call register if password and confirm password do not match", () => {
    // Arrange
    const register = jest.fn();
    const wrapper = shallow(<Register register={register} />);
    const username = "Test";
    const email = "Test@gmail.com";
    const password = "Test";
    const confirmPassword = "Test123";
    const setStateSpy = jest.spyOn(wrapper.instance(), "setState");

    // Act
    wrapper.setState({ username, email, password, confirmPassword });
    const form = wrapper.find("form");
    form.simulate("submit", {
      preventDefault: jest.fn()
    });

    // Assert
    expect(register).not.toBeCalled();
    expect(setStateSpy).toBeCalledWith({
      validationError: "Password and confirm password don't match"
    });
  });

  it("Should onChange happen handleChange should appropriately update the state", () => {
    // Arrange
    const wrapper = shallow(<Register />);
    const changeValue = "Test123";
    const changeEvent = { target: { value: changeValue, name: "username" } };
    const setStateSpy = jest.spyOn(wrapper.instance(), "setState");

    // Act
    const usernameInputField = wrapper.find(TextField).at(0);
    usernameInputField.simulate("change", changeEvent);

    // Assert
    expect(setStateSpy).toBeCalledWith({ username: changeValue });
    expect(wrapper.state()).toEqual({
      username: changeValue,
      password: "",
      confirmPassword: "",
      email: "",
      validationError: null
    });
  });
});
