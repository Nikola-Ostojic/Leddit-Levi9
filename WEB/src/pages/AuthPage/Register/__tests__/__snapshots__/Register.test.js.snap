// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register component tests Should match snapshot 1`] = `
<div>
  <h3>
    Register
  </h3>
  <form
    method="POST"
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Username"
      margin="normal"
      name="username"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Email"
      margin="normal"
      name="email"
      onChange={[Function]}
      required={true}
      type="email"
      value=""
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Password"
      margin="normal"
      name="password"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Confirm password"
      margin="normal"
      name="confirmPassword"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
    <br />
    <br />
    <WithStyles(ForwardRef(Button))
      color="primary"
      type="submit"
      variant="contained"
    >
      Register
    </WithStyles(ForwardRef(Button))>
  </form>
</div>
`;
